cli/dstack/cli/commands/run/ssh_tunnel.py

```diff
@@ -0,0 +1,50 @@
+import socket
+import subprocess
+from contextlib import closing
+from pathlib import Path
+from typing import Dict, List
+
+from dstack.core.job import Job
+
+
+def get_free_port() -> int:
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as s:
+        s.bind(("", 0))
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        return s.getsockname()[1]
+
+
+def allocate_local_ports(jobs: List[Job]) -> Dict[int, int]:
+    ports = {}
+    for job in jobs:
+        ws_logs_port = int(job.env.get("WS_LOGS_PORT"))
+        if ws_logs_port:
+            ports[ws_logs_port] = get_free_port()
+        for app_spec in job.app_specs:
```

Comment:
`job.app_specs` can be `None`. `dstack run python-version --remote` fails:

```
✗ dstack run python-version --remote                                                                               
 RUN          WORKFLOW        SUBMITTED  OWNER             STATUS     TAG  BACKENDS 
 red-eagle-1  python-version  now        Victor Skvortsov  Submitted       aws      

Provisioning... It may take up to a minute. ✓

To interrupt, press Ctrl+C.

Traceback (most recent call last):
  File "/Users/r4victor/Projects/dstack/venv/bin/dstack", line 33, in <module>
    sys.exit(load_entry_point('dstack', 'console_scripts', 'dstack')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/r4victor/Projects/dstack/dstack/cli/dstack/cli/main.py", line 50, in main
    args.func(args)
  File "/Users/r4victor/Projects/dstack/dstack/cli/dstack/cli/commands/__init__.py", line 41, in __command
    self._command(args)
  File "/Users/r4victor/Projects/dstack/dstack/cli/dstack/core/error.py", line 31, in decorator
    func(*args, **kwargs)
  File "/Users/r4victor/Projects/dstack/dstack/cli/dstack/core/error.py", line 41, in decorator
    func(*args, **kwargs)
  File "/Users/r4victor/Projects/dstack/dstack/cli/dstack/core/error.py", line 51, in decorator
    func(*args, **kwargs)
  File "/Users/r4victor/Projects/dstack/dstack/cli/dstack/cli/commands/run/__init__.py", line 332, in _command
    poll_run(repo_data, jobs, backend, ssh_key_private)
  File "/Users/r4victor/Projects/dstack/dstack/cli/dstack/cli/commands/run/__init__.py", line 208, in poll_run
    ports = allocate_local_ports(jobs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/r4victor/Projects/dstack/dstack/cli/dstack/cli/commands/run/ssh_tunnel.py", line 23, in allocate_local_ports
    for app_spec in job.app_specs:
TypeError: 'NoneType' object is not iterable
```

I personally don't think `job.app_specs` should be `Optional` and we should rewrite such values to be represented with empty lists. But until we do this we need to handle `None` values.

Reply:
Fixed